WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewModel.swift

```diff
@@ -771,6 +776,18 @@ private extension ProductFormViewModel {
     private func isNewTemplateProduct() -> Bool {
         originalProduct.productID == .zero && !originalProduct.isEmpty()
     }
+
+    /// Listen to storage changes and update the product when needed
+    /// Only needed for updates that are saved remotely on other screens
+    ///
+    private func configureEntityListener() {
+        entityListener.onUpsert = { [weak self] product in
+            guard let self = self else { return }
+            // Custom fields are updated on a different screen, update the product to reflect the last changes
+            resetProduct(EditableProductModel(product: originalProduct.product.copy(customFields: product.customFields)))
+            self.product = EditableProductModel(product: product.copy(customFields: product.customFields))
+        }
+    }
```

Comment:
After implementing the `MetaDataStore` changes, I noticed that the root issue was fixed for orders and not for products (when I go back then re-open the custom fields, the changes are not reflected), then I noticed that the Order Details screen uses an `EntityListener` to observe the storage changes of the order then [updates](https://github.com/woocommerce/woocommerce-ios/blob/685602e979a15339b7006b36bdd541dc756b3078/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/OrderDetailsViewController.swift#L210-L218) the `order` property, which is why the issue is not present there.

I tried to replicate this here instead of my previous attempt using the callback, I think this is cleaner, the only difference from what we do in order details is that I update only the customFields to preserve the other user changes (if any), I think this solution would work also to replace the `updateProductVariations` function, but it's another topic.

Please let me know what you think, if you think the callback function is better I can revert to using it.

Reply:
> I was concerned that this would cause other fields to be updated more than once but then I realized from [here](https://github.com/woocommerce/woocommerce-ios/blob/2077debb97b5020c3aa1faaece30961beabf73e8/WooCommerce/Classes/ViewRelated/Products/Edit%20Product/ProductFormViewModel.swift#L416-L446) that we were not updating storage from each individual screen for other fields. Instead, we use the Save button on the main product form to send all the changes to remote and storage.
>
> I believe there's such difference because we are reusing the same CustomFieldsListView with order, and in the order details screen we save each details individually.

Yes, this is a tradeoff we decided to accept to allow better reusability of the screen, and to decided to add a top banner to make it less confusing for the users.

> However, I'd prefer reverting back to the previous way of saving the parent product as you pointed out that only custom fields are using it.

Although I would love to know why you think that the previous way is better than using `EntityListener`, I applied the revert below to move forward with merging the PR.